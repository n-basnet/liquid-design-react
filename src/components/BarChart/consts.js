import COLORS from '~/utils/consts/colors'

export const DEFAULT_COLORS = [
  COLORS.VIBRANT_GREEN,
  COLORS.VIBRANT_MAGENTA,
  COLORS.VIBRANT_CYAN,
  COLORS.VIBRANT_YELLOW,
  COLORS.RICH_PURPLE,
  COLORS.RICH_BLUE,
  COLORS.RICH_GREEN,
]

export const BARS_WRAPPER_HEIGHT = 244

export const TICKS_OFFSET = 35
export const TICKS_PADDING = 50
export const TICKS_LINE_HEIGHT = 16
