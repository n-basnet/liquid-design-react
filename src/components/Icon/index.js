import React from 'react'
import PropTypes from 'prop-types'

import Favorite from '~/graphics/svgs/favorite.svg'

const ICONS = {
  favorite: Favorite,
}

const Icon = ({
  name,
  size = 20,
  style = {},
}) => {
  const SVGIconComponent = ICONS[name]
  return (
    <SVGIconComponent
      width={`${size}px`}
      height={`${size}px`}
      style={style}
    />
  )
}

Icon.propTypes = {
  name: PropTypes.string.isRequired,
  size: PropTypes.number,
  style: PropTypes.object,
}

export default Icon
