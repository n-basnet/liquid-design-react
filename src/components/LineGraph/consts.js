export const CHART_PIXEL_PADDING = {
  top: 75,
  bottom: 45,
  left: 110,
  right: 35,
}

export const LINE_CHART_LABELS_PRECISION = 0

export const CHART_PIXEL_WIDTH = 1000

export const DESKTOP_LINE_SIZE = 5
export const MOBILE_LINE_SIZE = 3

export const CHART_PIXEL_HEIGHT = 400

export const DATA_POINT_SIZE = 5

export const MOBILE_CHART_PIXEL_PADDING = {
  top: 45,
  bottom: 75,
  left: 75,
  right: 25,
}

export const TICK_COUNT = {
  X: 4,
  Y: 6,
}
export const MOBILE_CHART_HEIGHT_AMOUNT = 0.7
