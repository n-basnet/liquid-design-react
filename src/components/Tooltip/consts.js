export const ARROW_SIZE = 11

export const WALLS = {
  top: 'top',
  bottom: 'bottom',
  left: 'left',
  right: 'right',
}

export const SIDES = {
  left: 'left',
  right: 'right',
}

export const WALLS_KEYS = Object.keys(WALLS)
export const SIDES_KEYS = Object.keys(SIDES)
