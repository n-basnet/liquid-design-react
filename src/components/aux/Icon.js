import React from 'react'
import PropTypes from 'prop-types'

import Favorite from '~/graphics/svgs/favorite.svg'
import CircleX from '~/graphics/svgs/circle-x.svg'
import Star from '~/graphics/svgs/star.svg'
import StarHalf from '~/graphics/svgs/star-half.svg'
import Dot from '~/graphics/svgs/dot.svg'
import DotHalf from '~/graphics/svgs/dot-half.svg'

const ICONS = {
  favorite: Favorite,
  circleX: CircleX,
  star: Star,
  starHalf: StarHalf,
  dot: Dot,
  dotHalf: DotHalf,
}

const Icon = ({
  name,
  size = 14,
  style = {},
}) => {
  const SVGIconComponent = ICONS[name]
  return (
    <SVGIconComponent
      width={`${size}px`}
      height={`${size}px`}
      style={style}
    />
  )
}

Icon.propTypes = {
  name: PropTypes.string.isRequired,
  size: PropTypes.number,
  style: PropTypes.object,
}

export default Icon
