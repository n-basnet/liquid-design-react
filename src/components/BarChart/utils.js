import uniqid from 'uniqid'
import { find, pluck, head, prop, uniqBy } from 'ramda'

import { DEFAULT_COLORS } from '~/components/BarChart/consts'
import { times, getElementModular } from '~/utils/misc'
import { SCREEN_SIZES } from '~/utils/styling'

const getLabels = data =>
  data
    .reduce((labels, { values }) => {
      const valuesLabels = pluck('label', values)
      return uniqBy(prop('name'), [...labels, ...valuesLabels])
    }, [])
    .map((label, i) => ({
      id: `l-${uniqid()}`,
      color: getElementModular(i, DEFAULT_COLORS),
      ...label,
    }))

const getLargestValue = data => {
  const allValues = pluck('values', data).map(pluck('value'))
  return head(allValues.reduce((a, b) => [...a, ...b], []).sort((a, b) => b - a))
}

const findByName = (name, array) => find(element => element.name === name, array)

const roundToThousands = num => Math.round(num / 1000) * 1000
export const getTicks = (data, offset = 0, amount = 6) => {
  if (data.length === 0) {
    return []
  }
  const maximum = (100 * getLargestValue(data)) / (100 - offset)
  const tickValue = roundToThousands(maximum / amount)
  return times(amount + 1).map(i => i * tickValue)
}

export const recomputeData = data => {
  const largestValue = getLargestValue(data)
  const labels = getLabels(data)

  return {
    data: data.map((point, i) => ({
      ...point,
      id: `p-${uniqid()}`,
      values: point.values.map(value => ({
        id: `v-${uniqid()}`,
        valueAsPercentage: (value.value * 100) / largestValue,
        ...value,
        label: findByName(value.label.name, labels),
      })),
    })),
    labels,
  }
}

export const isMobile = () => window.innerWidth <= SCREEN_SIZES.phone
