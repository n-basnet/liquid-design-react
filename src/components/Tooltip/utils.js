import { css } from 'styled-components'

import { ARROW_SIZE, WALLS, SIDES } from '../../components/Tooltip/consts'

const isHorizontalWall = wall => wall === WALLS.top || wall === WALLS.bottom

const getYAttribute = ({ wall, side }) => {
  if (
    wall === WALLS.top ||
    (wall === WALLS.left && side === SIDES.right) ||
    (wall === WALLS.right && side === SIDES.left)
  ) {
    return WALLS.top
  } else {
    return WALLS.bottom
  }
}

const getXAttribute = ({ wall, side }) => {
  if (
    wall === WALLS.right ||
    (wall === WALLS.top && side === SIDES.right) ||
    (wall === WALLS.bottom && side === SIDES.left)
  ) {
    return SIDES.right
  } else {
    return SIDES.left
  }
}

const getArrow = ({ color, direction, size = ARROW_SIZE }) =>
  ({
    top: css`
      border-bottom: ${size}px solid ${color};
      border-left: ${size}px solid transparent;
      border-right: ${size}px solid transparent;
    `,
    bottom: css`
      border-top: ${size}px solid ${color};
      border-left: ${size}px solid transparent;
      border-right: ${size}px solid transparent;
    `,
    right: css`
      border-top: ${size}px solid transparent;
      border-bottom: ${size}px solid transparent;
      border-left: ${size}px solid ${color};
    `,
    left: css`
      border-top: ${size}px solid transparent;
      border-bottom: ${size}px solid transparent;
      border-right: ${size}px solid ${color};
    `,
  }[direction])

export const getArrowStyle = ({ color, wall, side }) => css`
  ${getYAttribute({ wall, side })}: ${isHorizontalWall(wall) ? -10 : 20}px;
  ${getXAttribute({ wall, side })}: ${isHorizontalWall(wall)
    ? ARROW_SIZE
    : -ARROW_SIZE}px;
  ${getArrow({ direction: wall, color })};
`

const getYPosValue = ({ wall, side }) => {
  if (isHorizontalWall(wall)) {
    return wall === WALLS.top ? 28 : 43
  } else {
    return wall === side ? -17 : -26
  }
}

const getXPosAttribute = ({ wall, side }) => {
  if (
    wall === WALLS.left ||
    (wall === WALLS.top && side === SIDES.left) ||
    (wall === WALLS.bottom && side === SIDES.right)
  ) {
    return SIDES.left
  } else {
    return SIDES.right
  }
}

export const getPosition = props => css`
  ${getYAttribute(props)}: ${getYPosValue(props)}px;
  ${getXPosAttribute(props)}: ${isHorizontalWall(props.wall) ? -10 : 40}px;
`
