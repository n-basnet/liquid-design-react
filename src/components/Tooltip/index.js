import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import enhanceWithClickOutside from 'react-click-outside'
import cx from 'classnames'

import { Glyph, ICON_CLASSNAME } from '~/elements/Icon'
import { getPosition, getArrowStyle } from '~/components/Tooltip/utils'
import { WALLS, WALLS_KEYS, SIDES, SIDES_KEYS } from '~/components/Tooltip/consts'
import { hasCSSFilters } from '~/utils/featureDetects'
import { getClassName } from '~/components/aux/hoc/attachClassName'

export const TOOLTIP_WRAPPER_CLASSNAME = getClassName({ name: 'TooltipWrapper' })

const getIconColor = props => (props.isOpen ? 'darker' : props.isHovered ? 'dark' : 'base')

const TooltipWrapper = styled.span`
  position: relative;
  ${props => css`
    > .${ICON_CLASSNAME} svg {
      fill: ${props.theme.colors.primary[getIconColor(props)]};
    }
  `};
`

const TooltipContentWrapper = styled.div`
  position: absolute;
  padding: 19px 25px;
  width: 250px;
  ${props => css`
    ${getPosition(props)};
    z-index: ${props.theme.zIndex.tooltips};
    background-color: ${props.theme.colors.white.base};
    border-radius: ${props.theme.borderRadius};
    ${hasCSSFilters()
    ? css`
          filter: drop-shadow(0 10px 10px rgba(0, 0, 0, 0.1));
        `
    : css`
          box-shadow: ${props.theme.boxShadow};
        `};

    ${hasCSSFilters() &&
      css`
        &:after {
          content: '';
          position: absolute;
          width: 0;
          height: 0;
          ${props => css`
            ${getArrowStyle({
    color: props.theme.colors.white.base,
    wall: props.wall,
    side: props.side,
  })};
          `};
        }
      `};
  `};
`

export class Tooltip extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
    /** the wall on which the arrow will be displayed */
    wall: PropTypes.oneOf(WALLS_KEYS),
    /** the side of the wall - left or right - on which the arrow should be displayed */
    side: PropTypes.oneOf(SIDES_KEYS),
    /** object for overriding the style of the Tooltip content */
    contentStyle: PropTypes.object,
    /** force open state for tooltip */
    isOpen: PropTypes.bool,
    /** provide a custom trigger element, instead of the default info icon */
    customTrigger: PropTypes.func,
    /** callback on toggling the tooltip */
    onToggle: PropTypes.func,
    className: PropTypes.string,
  }
  static defaultProps = {
    wall: WALLS.top,
    side: SIDES.left,
    contentStyle: {},
    customTrigger: undefined,
    isOpen: null,
    onToggle: undefined,
    className: null,
  }
  state = {
    isOpen: false,
  }
  toggle = () => {
    this.setState(
      ({ isOpen }) => ({ isOpen: !isOpen }),
      () => {
        this.props.onToggle && this.props.onToggle(this.state.isOpen)
      }
    )
  }
  /** setTimeout hack is to call toggle function firstly instead of handleMouseEnter on safari ios **/
  handleMouseEnter = () => setTimeout(() => this.setState({ isHovered: true }), 0)
  setHoverToFalse = () => this.setState({ isHovered: false })
  handleClickOutside = () => {
    this.setState({ isOpen: false })
    this.props.onToggle && this.props.onToggle(false)
  }
  render() {
    const { isHovered } = this.state
    const { children, wall, side, contentStyle, customTrigger, className, ...props } = this.props
    const isTooltipOpen = this.props.isOpen !== null ? this.props.isOpen : this.state.isOpen
    const iconName = `tooltip${isHovered || isTooltipOpen ? 'Filled' : 'Empty'}`
    return (
      <TooltipWrapper
        isOpen={isTooltipOpen}
        isHovered={isHovered}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.setHoverToFalse}
        onTouchStart={this.setHoverToFalse}
        className={cx(getClassName(Tooltip), className)}
        {...props}
      >
        {customTrigger ? (
          customTrigger(this.toggle)
        ) : (
          <Glyph name={iconName} onClick={this.toggle} />
        )}
        {isTooltipOpen && (
          <TooltipContentWrapper
            className={TOOLTIP_WRAPPER_CLASSNAME}
            style={contentStyle}
            wall={wall}
            side={side}
          >
            {children}
          </TooltipContentWrapper>
        )}
      </TooltipWrapper>
    )
  }
}

export default enhanceWithClickOutside(Tooltip)
