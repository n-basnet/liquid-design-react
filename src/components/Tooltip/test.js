import { ICON_CLASSNAME } from '../../elements/Icon'
import Tooltip from '.'
import { getArrowStyle, getPosition } from './utils'
import { WALLS_KEYS, SIDES_KEYS } from './consts'
import { everyComponentTestSuite, getWrapper } from '../../utils/testUtils'

describe('Tooltip', () => {
  const tooltipContent = 'some content'
  const getTooltipWrapper = getWrapper(Tooltip, { children: tooltipContent })

  it('displays and hides content on click', () => {
    const wrapper = getTooltipWrapper()
    const TooltipIcon = wrapper.find(`.${ICON_CLASSNAME}`).first()
    const getFirstDivText = () =>
      wrapper
        .find('div')
        .first()
        .text()
    expect(getFirstDivText()).toEqual('')
    TooltipIcon.simulate('click')
    expect(getFirstDivText()).toEqual(tooltipContent)
    TooltipIcon.simulate('click')
    expect(getFirstDivText()).toEqual('')
  })

  it('utils - getArrowStyle', () => {
    WALLS_KEYS.map(wall =>
      SIDES_KEYS.map(side => {
        expect(getArrowStyle({ color: 'white', wall, side })).toMatchSnapshot()
      }),
    )
  })

  it('utils - getPosition', () => {
    WALLS_KEYS.map(wall =>
      SIDES_KEYS.map(side => {
        expect(getPosition({ wall, side })).toMatchSnapshot()
      }),
    )
  })

  everyComponentTestSuite(getTooltipWrapper, Tooltip, 'Tooltip')
})
