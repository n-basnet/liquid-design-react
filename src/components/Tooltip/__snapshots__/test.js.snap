// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip utils - getArrowStyle 1`] = `
Array [
  "
  ",
  "top",
  ": ",
  "-10",
  "px;
  ",
  "left",
  ": ",
  "11",
  "px;
  ",
  "
      border-bottom: ",
  "11",
  "px solid ",
  "white",
  ";
      border-left: ",
  "11",
  "px solid transparent;
      border-right: ",
  "11",
  "px solid transparent;
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 2`] = `
Array [
  "
  ",
  "top",
  ": ",
  "-10",
  "px;
  ",
  "right",
  ": ",
  "11",
  "px;
  ",
  "
      border-bottom: ",
  "11",
  "px solid ",
  "white",
  ";
      border-left: ",
  "11",
  "px solid transparent;
      border-right: ",
  "11",
  "px solid transparent;
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 3`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "-10",
  "px;
  ",
  "right",
  ": ",
  "11",
  "px;
  ",
  "
      border-top: ",
  "11",
  "px solid ",
  "white",
  ";
      border-left: ",
  "11",
  "px solid transparent;
      border-right: ",
  "11",
  "px solid transparent;
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 4`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "-10",
  "px;
  ",
  "left",
  ": ",
  "11",
  "px;
  ",
  "
      border-top: ",
  "11",
  "px solid ",
  "white",
  ";
      border-left: ",
  "11",
  "px solid transparent;
      border-right: ",
  "11",
  "px solid transparent;
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 5`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "20",
  "px;
  ",
  "left",
  ": ",
  "-11",
  "px;
  ",
  "
      border-top: ",
  "11",
  "px solid transparent;
      border-bottom: ",
  "11",
  "px solid transparent;
      border-right: ",
  "11",
  "px solid ",
  "white",
  ";
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 6`] = `
Array [
  "
  ",
  "top",
  ": ",
  "20",
  "px;
  ",
  "left",
  ": ",
  "-11",
  "px;
  ",
  "
      border-top: ",
  "11",
  "px solid transparent;
      border-bottom: ",
  "11",
  "px solid transparent;
      border-right: ",
  "11",
  "px solid ",
  "white",
  ";
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 7`] = `
Array [
  "
  ",
  "top",
  ": ",
  "20",
  "px;
  ",
  "right",
  ": ",
  "-11",
  "px;
  ",
  "
      border-top: ",
  "11",
  "px solid transparent;
      border-bottom: ",
  "11",
  "px solid transparent;
      border-left: ",
  "11",
  "px solid ",
  "white",
  ";
    ",
  ";
",
]
`;

exports[`Tooltip utils - getArrowStyle 8`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "20",
  "px;
  ",
  "right",
  ": ",
  "-11",
  "px;
  ",
  "
      border-top: ",
  "11",
  "px solid transparent;
      border-bottom: ",
  "11",
  "px solid transparent;
      border-left: ",
  "11",
  "px solid ",
  "white",
  ";
    ",
  ";
",
]
`;

exports[`Tooltip utils - getPosition 1`] = `
Array [
  "
  ",
  "top",
  ": ",
  "28",
  "px;
  ",
  "left",
  ": ",
  "-10",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 2`] = `
Array [
  "
  ",
  "top",
  ": ",
  "28",
  "px;
  ",
  "right",
  ": ",
  "-10",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 3`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "43",
  "px;
  ",
  "right",
  ": ",
  "-10",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 4`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "43",
  "px;
  ",
  "left",
  ": ",
  "-10",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 5`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "-17",
  "px;
  ",
  "left",
  ": ",
  "40",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 6`] = `
Array [
  "
  ",
  "top",
  ": ",
  "-26",
  "px;
  ",
  "left",
  ": ",
  "40",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 7`] = `
Array [
  "
  ",
  "top",
  ": ",
  "-26",
  "px;
  ",
  "right",
  ": ",
  "40",
  "px;
",
]
`;

exports[`Tooltip utils - getPosition 8`] = `
Array [
  "
  ",
  "bottom",
  ": ",
  "-17",
  "px;
  ",
  "right",
  ": ",
  "40",
  "px;
",
]
`;
